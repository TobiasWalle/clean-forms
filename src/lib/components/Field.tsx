import React, { memo, useCallback, useEffect, useLayoutEffect, useRef } from 'react';
import { FormApi } from '../api';
import { useFieldGroupContext } from '../contexts/field-group-context';
import { useFormContext } from '../contexts/form-context';
import { FieldStatus } from '../statusTracking';
import { createPath, Path } from '../utils';
import { FieldError } from '../validation';

export type FieldId = string;

export interface InputProps<Value> extends FieldStatus {
  name?: string;
  value: Value;
  onFocus: () => void;
  onBlur: () => void;
  onChange: (value: Value) => void;
  error: FieldError;
  valid: boolean;
  inValid: boolean;
}

export interface InnerFieldProps<Value, Model = any, CustomProps = {}> {
  input: InputProps<Value>;
  form: FormApi<Model>;
  custom: CustomProps;
}

export type FieldRenderFunction<Value = any, Model = any, RenderProps = {}> =
  React.FunctionComponent<InnerFieldProps<Value, Model, RenderProps>>;

export interface FieldPropsWithoutRender {
  name: string | null;
}

export type FieldProps<Value, CustomProps = {}> = FieldPropsWithoutRender & {
  render: FieldRenderFunction<Value, any, CustomProps>;
  inner: CustomProps;
};

function _Field<Value = any, CustomProps = {}>({
  name,
  render,
  inner: custom
}: FieldProps<Value, CustomProps>) {
  const { form, ...formContext } = useFormContext();
  const groupContext = useFieldGroupContext();

  const path: Path = createPath(groupContext.path, name);
  const fieldId: FieldId = createPath(groupContext.namespace, name);

  const value = form.getFieldValue(path);
  const status = form.getFieldStatus(fieldId);
  const error = form.getFieldError(path);

  useLayoutEffect(() => {
    formContext.onFieldMount(fieldId);
    return () => formContext.onFieldUnmount(fieldId);
  }, [fieldId]);

  const handleFocus = useCallback(() => {
    formContext.onFieldFocus(fieldId);
  }, [formContext.onFieldFocus, fieldId]);

  const handleBlur = useCallback(() => {
    formContext.onFieldBlur(fieldId);
  }, [formContext.onFieldBlur, fieldId]);

  const handleChange = useCallback((newValue: Value) => {
    formContext.onFieldChange(fieldId, path, newValue);
  }, [formContext.onFieldBlur, fieldId, path]);

  const input: InputProps<Value> = {
    name: name || undefined,
    value,
    onFocus: handleFocus,
    onBlur: handleBlur,
    onChange: handleChange,
    ...status,
    error,
    valid: !error,
    inValid: !!error,
  };

  return render({ input, custom, form });
}

export const Field = memo(_Field);
