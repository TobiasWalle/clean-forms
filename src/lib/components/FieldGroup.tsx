import * as React from 'react';
import { FieldGroupContext, useFieldGroupContext } from '../contexts/field-group-context';
import { createPath } from '../utils';

export interface FieldGroupProps {
  name: string;
  accessor?: string;
  children?: React.ReactNode;
}

function _FieldGroup({ name, accessor, children }: FieldGroupProps) {
  const parentContext = useFieldGroupContext();
  const newContext = {
    path: createPath(parentContext.path, accessor || name),
    namespace: createPath(parentContext.namespace, name)
  };
  return (
    <FieldGroupContext.Provider value={newContext}>
      {children}
    </FieldGroupContext.Provider>
  );
}

export const FieldGroup = React.memo(_FieldGroup);
