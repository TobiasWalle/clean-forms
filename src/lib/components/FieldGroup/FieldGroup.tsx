import * as PropTypes from 'prop-types';
import * as React from 'react';
import { createPath } from '../../utils/createPath';

export const fieldGroupContextTypes = {
  groups: PropTypes.arrayOf(PropTypes.string),
};

export interface FieldGroupContext {
  groups?: string[];
}

export interface FieldGroupProps {
  name: string;
}

export interface FieldGroupState {
}

export class FieldGroup extends React.Component<FieldGroupProps, FieldGroupState> {
  public static childContextTypes = fieldGroupContextTypes;
  public static contextTypes = fieldGroupContextTypes;
  public context: FieldGroupContext;

  public render() {
    return this.props.children;
  }

  public getChildContext(): FieldGroupContext {
    return {
      groups: createPath(this.context.groups, this.props.name)
    };
  }
}
