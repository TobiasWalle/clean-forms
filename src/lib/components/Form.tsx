import * as React from 'react';
import {
  FormEvent,
  forwardRef,
  memo,
  MutableRefObject,
  Ref,
  RefAttributes,
  useCallback,
  useEffect,
  useImperativeHandle,
  useLayoutEffect,
  useMemo,
  useRef,
  useState,
} from 'react';
import { FieldContext, FieldContextValue } from '../contexts/fieldContext';
import {
  FormContext,
  FormContextValue,
  OnFieldBlur,
  OnFieldChange,
  OnFieldMount,
  OnFieldUnmount,
  SetArrayGetKey,
} from '../contexts/formContext';
import { useMemorizedPath } from '../hooks';
import { useAsRef } from '../hooks/useAsRef';
import { useControllableState } from '../hooks/useControllableState';
import { useFormReadApi } from '../hooks/useFormReadApi';
import { useShallowMemo } from '../hooks/useShallowMemo';
import {
  fieldPath as createFieldPath,
  getPathAsString,
  Path,
  path as createPath,
} from '../models';
import { cloneFieldStatus, FieldStatusMapping } from '../statusTracking';
import { FieldStatusUpdater } from '../statusTracking/FieldStatusUpdater';
import { FieldRegister, FieldRegisterChanges } from '../utils';
import { StateUpdater } from '../utils/StateUpdater';
import {
  FieldErrorMapping,
  validateModel,
  ValidationDefinition,
} from '../validation';

import { GetKey } from './FieldArrayItems';

export type OnChange<Model> = (value: Model) => void;
export type OnStatusMappingChange = (mapping: FieldStatusMapping) => void;
export type OnSubmit = (event?: FormEvent<HTMLFormElement>) => void;
export type OnErrorsChange = (errors: FieldErrorMapping) => void;

export interface FormRef {
  /** Trigger a form submit */
  submit: () => void;
}

export type FormProps<Model> = {
  value: Model;
  onChange?: OnChange<Model>;

  status?: FieldStatusMapping;
  onStatusChange?: OnStatusMappingChange;

  onSubmit?: OnSubmit;
  onValidSubmit?: OnSubmit;
  onInValidSubmit?: OnSubmit;
  onErrorsChange?: OnErrorsChange;
  validation?: ValidationDefinition<Model>;
  formProps?: JSX.IntrinsicElements['form'];

  /** Strictly check if the name structure matches the form model. Default is true. */
  strict?: boolean;

  children?: React.ReactNode;
} & RefAttributes<FormRef>;

function useIsMounted(): Readonly<MutableRefObject<boolean>> {
  const mounted = useRef(false);
  useLayoutEffect(() => {
    mounted.current = true;
    return () => {
      mounted.current = false;
    };
  }, []);
  return mounted;
}

function _Form<Model = any>(props: FormProps<Model>, ref: Ref<FormRef>) {
  const {
    value,
    onStatusChange,
    validation,
    formProps,
    onChange,
    onSubmit,
    onErrorsChange,
    onInValidSubmit,
    onValidSubmit,
    strict,
  } = props;
  const [status, setStatus] = useControllableState<FieldStatusMapping>({
    initialStateIfNotControlled: {},
    onChange: onStatusChange,
    controlledState: props.status,
  });

  const setValue = useCallback(
    (newModel: Model) => onChange && onChange(newModel),
    [onChange]
  );

  const valueRef = useAsRef(value);
  const valueUpdaterRef = useRef(
    new StateUpdater<Model>(() => valueRef.current, () => isMountedRef.current)
  );
  useEffect(() => {
    valueUpdaterRef.current.registerOnChange(onChange);
  }, [onChange]);

  const statusRef = useAsRef(status);
  const statusUpdaterRef = useRef(
    new StateUpdater<FieldStatusMapping>(
      () => statusRef.current,
      () => isMountedRef.current
    )
  );
  useEffect(() => {
    statusUpdaterRef.current.registerOnChange(setStatus);
  }, [setStatus]);

  // ArrayGetKeyMapping
  const arrayGetKeyMappingRef = useRef<Map<string, GetKey<any>>>(new Map());

  const setArrayGetKey: SetArrayGetKey<Model> = useCallback((path, getKey) => {
    arrayGetKeyMappingRef.current.set(getPathAsString(path), getKey);
  }, []);

  const removeArrayGetKeyFunction = useCallback((path: Path<Model>) => {
    arrayGetKeyMappingRef.current.delete(getPathAsString(path));
  }, []);

  // FieldRegisterRef
  const handleFieldRegisterChanges = useCallback(
    (changes: FieldRegisterChanges): void => {
      let updatedStatus = status;
      changes.registered.forEach(newPath => {
        updatedStatus = fieldStatusUpdaterRef.current.addIfFieldNotExists(
          updatedStatus,
          newPath
        );
      });

      changes.unregistered.forEach(removedPath => {
        updatedStatus = fieldStatusUpdaterRef.current.removeIfFieldExists(
          updatedStatus,
          removedPath
        );
        removeArrayGetKeyFunction(removedPath);
      });

      statusUpdaterRef.current.update(() => updatedStatus);
    },
    [status, removeArrayGetKeyFunction]
  );

  const [fieldRegister] = useState(new FieldRegister());
  useLayoutEffect(() => {
    fieldRegister.addListener(handleFieldRegisterChanges);
    return () => fieldRegister.removeListener(handleFieldRegisterChanges);
  }, [fieldRegister, handleFieldRegisterChanges]);
  // FieldRegisterRef END

  const fieldStatusUpdaterRef = useRef(new FieldStatusUpdater(fieldRegister));

  const isMountedRef = useIsMounted();

  // API
  const fieldErrorMapping = useMemo<FieldErrorMapping>(
    () =>
      validation
        ? validateModel({ model: value, validationDefinition: validation })
        : {},
    [value, validation]
  );

  useEffect(() => {
    onErrorsChange && onErrorsChange(fieldErrorMapping);
  }, [fieldErrorMapping, onErrorsChange]);

  const {
    valid,
    invalid,
    getFieldStatus,
    getFieldError,
    getFieldValue,
  } = useFormReadApi({
    value,
    status,
    validationDefinition: validation,
    fieldErrorMapping,
    strict,
  });

  // Callbacks
  const handleFieldMount: OnFieldMount<Model> = useCallback(
    path => {
      fieldRegister.register(path);
    },
    [fieldRegister]
  );

  const handleFieldUnmount: OnFieldUnmount<Model> = useCallback(
    path => {
      if (!isMountedRef.current) {
        // Cancel if the form is not mounted anymore
        return;
      }
      fieldRegister.unregister(path);
    },
    [fieldRegister, isMountedRef]
  );

  const getFieldStatusRef = useAsRef(getFieldStatus);
  const handleFieldBlur: OnFieldBlur<Model> = useCallback(
    path => {
      const newStatus = cloneFieldStatus(getFieldStatusRef.current(path), {
        touched: true,
      });
      statusUpdaterRef.current.update(oldStatus => ({
        ...oldStatus,
        [getPathAsString(path)]: newStatus,
      }));
    },
    [getFieldStatusRef]
  );

  const handleFieldChange: OnFieldChange<Model> = useCallback(
    (id, path, fieldValue) => {
      valueUpdaterRef.current.updateDeep(getPathAsString(path), fieldValue);

      const updatedStatus = fieldStatusUpdaterRef.current.markAsDirty(
        statusRef.current,
        id
      );

      statusUpdaterRef.current.update(() => updatedStatus);
    },
    [statusRef]
  );
  // Callbacks End

  const markAllAsTouched = useCallback(() => {
    const updatedStatus = fieldStatusUpdaterRef.current.markAllAsTouched(
      statusRef.current
    );
    statusUpdaterRef.current.update(() => updatedStatus);
  }, [statusRef]);

  const submit = useCallback(
    (event?: React.FormEvent<HTMLFormElement>) => {
      markAllAsTouched();
      onSubmit && onSubmit(event);
      if (valid) {
        onValidSubmit && onValidSubmit(event);
      } else {
        onInValidSubmit && onInValidSubmit(event);
      }
    },
    [markAllAsTouched, onSubmit, valid, onValidSubmit, onInValidSubmit]
  );

  const handleSubmit = useCallback(
    (event: React.FormEvent<HTMLFormElement>) => {
      event.stopPropagation();
      event.preventDefault();

      formProps && formProps.onSubmit && formProps.onSubmit(event);

      submit(event);
    },
    [submit, formProps]
  );

  const formContext: FormContextValue<Model> = useShallowMemo({
    valid,
    invalid,
    getFieldStatus,
    getFieldError,
    getFieldValue,
    onFieldBlur: handleFieldBlur,
    onFieldChange: handleFieldChange,
    onFieldMount: handleFieldMount,
    onFieldUnmount: handleFieldUnmount,
    setArrayGetKey,
  });

  const error = getFieldError();
  const markAsTouched = useCallback(() => handleFieldBlur(createPath()), [
    handleFieldBlur,
  ]);

  const rootFieldContext: FieldContextValue<Model> = useShallowMemo({
    fieldPath: useMemorizedPath(createFieldPath<Model>()),
    modelPath: useMemorizedPath(createPath<Model>()),
    name: '',
    value,
    setValue,
    markAsTouched,
    error,
    valid: !error,
    invalid: !!error,
    ...getFieldStatus(),
  });

  useImperativeHandle(ref, () => ({ submit }), [submit]);

  return (
    <FormContext.Provider value={formContext}>
      <FieldContext.Provider value={rootFieldContext}>
        <form {...formProps} onSubmit={handleSubmit}>
          {props.children}
        </form>
      </FieldContext.Provider>
    </FormContext.Provider>
  );
}

export type Form = typeof _Form;
export const Form: Form = memo(forwardRef(_Form)) as any;
