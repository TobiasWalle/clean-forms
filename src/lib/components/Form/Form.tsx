import * as PropTypes from 'prop-types';
import * as React from 'react';
import { updateDeep } from '../../utils';
import { Errors, ValidationDefinition, Validator } from '../../validation';

export type OnValueChange<Model> = (path: string[], value: any) => void;

export interface FormContext<Model> {
  model: Model;
  onValueChange: OnValueChange<Model>;
}

export const formContextTypes = {
  model: PropTypes.object,
  onValueChange: PropTypes.func,
};

export interface FormMeta<Model> {
  errors: Errors<Model> | undefined;
}

export interface FormProps<Model, FormValidation extends ValidationDefinition<Model> = ValidationDefinition<Model>> {
  model: Model;
  validation?: FormValidation;
  onChange?: (model: Model, meta: FormMeta<Model>) => void;
}

export class Form<Model = any, FormValidation extends ValidationDefinition<Model> = any> extends React.Component<FormProps<Model, FormValidation>, {}> {
  public static childContextTypes = formContextTypes;

  public render() {
    return (
      <form>
        {this.props.children}
      </form>
    );
  }

  public getChildContext(): FormContext<Model> {
    return {
      model: this.props.model,
      onValueChange: this.onValueChange
    };
  }

  private onValueChange: OnValueChange<Model> = (name, value) => {
    this.updateModel(name, value);
  }

  private updateModel: OnValueChange<Model> = (path, value) => {
    const { onChange, model } = this.props;
    const newModel: Model = updateDeep(model, path, value);

    const meta = this.getMetaData(newModel);
    onChange && onChange(newModel, meta);
  }

  private getMetaData(model: Model): FormMeta<Model> {
    return {
      errors: this.validate(model)
    };
  }

  private validate(model: Model): Errors<Model> | undefined {
    const {
      validation = {}
    } = this.props;
    const validator = new Validator<Model>();
    return validator.validate(model, validation);
  }
}
