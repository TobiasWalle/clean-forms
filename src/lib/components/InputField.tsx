import * as React from 'react';
import { createField } from '../hocs';
import { getInputProps } from '../utils/getInputProps';

export const InputField = createField<string, JSX.IntrinsicElements['input']>(
  ({ input: fieldInputProps, custom }) => {
    return <input {...getInputProps(fieldInputProps, custom)} />;
  }
);
