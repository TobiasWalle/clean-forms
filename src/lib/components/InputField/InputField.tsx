import * as React from 'react';
import { createField } from '../../hocs';
import { FieldPropsWithoutRender, FieldRenderFunction, } from '../Field/Field';

export type InputValue = number | string;

export interface InputProps extends React.InputHTMLAttributes<HTMLInputElement> {
  label?: string;
}

export const renderInput: FieldRenderFunction<InputValue, InputProps> = ({ input: { name, value, onChange }, custom }) => {
  const label = custom && custom.label;
  return <label>
    <div>{label}</div>
    <input {...custom} name={name} value={value} onChange={event => onChange(event.target.value)}/>
  </label>;
};

export const InputField = createField<InputValue, InputProps>(renderInput);
