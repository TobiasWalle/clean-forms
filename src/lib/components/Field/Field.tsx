import * as PropTypes from 'prop-types';
import * as React from 'react';
import { FormApi } from '../../api/FormApi';
import { createPath } from '../../utils/createPath';
import { Path } from '../../utils/FieldRegister';
import { FieldStatus } from '../../utils/statusTracking/FieldStatus';
import { FieldError } from '../../utils/validation';
import { FieldGroupContext, fieldGroupContextTypes } from '../FieldGroup/FieldGroup';
import { FormContext, formContextTypes } from '../Form/Form';

export type FieldId = string;

export interface InputProps<Value> extends FieldStatus {
  name?: string;
  value: Value;
  onFocus: () => void;
  onBlur: () => void;
  onChange: (value: Value) => void;
  error: FieldError;
  valid: boolean;
  inValid: boolean;
}

export interface InnerFieldProps<Value, Model = any, CustomProps = {}> {
  input: InputProps<Value>;
  form: FormApi<Model>;
  custom: CustomProps;
}

export type FieldRenderFunction<Value = any, RenderProps = {}> =
  React.StatelessComponent<InnerFieldProps<Value, any, RenderProps>>;

export interface FieldPropsWithoutRender {
  name: string | null;
}

export type FieldProps<Value, CustomProps = {}> = FieldPropsWithoutRender & {
  render: FieldRenderFunction<Value, CustomProps>;
  inner: CustomProps;
};

export class Field<Value = any, CustomProps = any> extends React.Component<FieldProps<Value, CustomProps>, {}> {
  public static contextTypes = {
    ...formContextTypes,
    ...fieldGroupContextTypes
  };
  public context: FormContext<any> & FieldGroupContext;

  private fieldId: FieldId;
  private path: Path;

  public render() {
    const { name, render, inner: custom } = this.props;
    const { form } = this.context;
    this.updatePathAndId();

    const value = form.getFieldValue(this.path);
    const status = form.getFieldStatus(this.fieldId);
    const error = form.getFieldError(this.fieldId);
    const input: InputProps<Value> = {
      name: name || undefined,
      value,
      onFocus: this.onFocus,
      onBlur: this.onBlur,
      onChange: this.onChange,
      ...status,
      error,
      valid: !error,
      inValid: !!error,
    };

    return render({ input, custom, form });
  }

  private updatePathAndId(): void {
    this.path = createPath(this.context.path, this.props.name);
    this.fieldId = createPath(this.context.namespace, this.props.name);
  }

  public componentDidMount() {
    this.context.onFieldMount(this.fieldId);
  }

  public componentWillUnmount() {
    this.context.onFieldUnmount(this.fieldId);
  }

  private onFocus = () => {
    this.context.onFieldFocus(this.fieldId);
  };

  private onBlur = () => {
    this.context.onFieldBlur(this.fieldId);
  };

  private onChange = (value: any) => {
    this.context.onFieldChange(this.fieldId, this.path, value);
  };
}
