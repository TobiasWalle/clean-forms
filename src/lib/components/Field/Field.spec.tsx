import { mount, shallow } from 'enzyme';
import * as React from 'react';
import { mockEvent } from '../../../testUtils/mockEvent';
import { mockFormContext } from '../../../testUtils/mockFormContext';
import { FieldGroupContext } from '../FieldGroup/FieldGroup';
import { FormContext } from '../Form/Form';
import { renderInput } from '../InputField/InputField';
import { Field, InnerFieldProps } from './Field';

describe('Field', () => {
  it('should render', () => {
    const field = <Field name={'name'} render={renderInput}/>;
    const model = { name: 'value' };
    const context: FormContext<any> = mockFormContext(model);
    expect(shallow(field, { context })).toMatchSnapshot();
  });

  it('should pass props', () => {
    const mockRender = jest.fn(renderInput);
    const name = 'name';
    const value = 'value';
    const type = 'number';
    const field = <Field name={name} render={mockRender} inner={{ type }}/>;
    const model = { [name]: value };
    const context = mockFormContext(model);

    mount(field, { context });

    const expectedInput: InnerFieldProps<any, any> = {
      input: {
        name, value,
        onChange: expect.any(Function),
        onFocus: expect.any(Function),
        onBlur: expect.any(Function),
      },
      custom: { type },
      form: context.form
    };
    expect(mockRender).toBeCalledWith(expectedInput);
  });

  it('should trigger onFieldChange', () => {
    const name = 'name';
    const value = 'value';
    const onFieldChange = jest.fn();
    const field = <Field name={name} render={renderInput}/>;
    const model = { [name]: value };
    const context = mockFormContext(model, { onFieldChange });

    const element = mount(field, { context });
    const wrapper = element.find('input');
    const onChange = wrapper.props().onChange;

    expect(onFieldChange).not.toHaveBeenCalled();

    const newValue = 'newValue';
    onChange!(mockEvent(newValue));

    expect(onFieldChange).toHaveBeenCalledWith([name], newValue);
  });

  it('should trigger onFocus and onBlur', () => {
    const name = 'name';
    const value = 'value';
    const onFieldFocus = jest.fn();
    const onFieldBlur = jest.fn();
    const field = <Field name={name} render={renderInput}/>;
    const model = { [name]: value };
    const context = mockFormContext(model, { onFieldFocus, onFieldBlur });

    const element = mount(field, { context });
    const wrapper = element.find('input');
    const onFocus = wrapper.props().onFocus;
    const onBlur = wrapper.props().onBlur;

    expect(onFieldFocus).not.toHaveBeenCalled();
    expect(onFieldBlur).not.toHaveBeenCalled();

    onFocus!({} as any);

    expect(onFieldFocus).toHaveBeenCalled();
    expect(onFieldBlur).not.toHaveBeenCalled();

    onBlur!({} as any);

    expect(onFieldBlur).toHaveBeenCalled();
  });

  it('should trigger onMount and onUnmount', () => {
    const name = 'name';
    const value = 'value';
    const onFieldMount = jest.fn();
    const onFieldUnmount = jest.fn();
    const field = <Field name={name} render={renderInput}/>;
    const model = { [name]: value };
    const context = mockFormContext(model, { onFieldMount, onFieldUnmount });

    const element = mount(field, { context });

    expect(onFieldMount).toHaveBeenCalled();
    expect(onFieldUnmount).not.toHaveBeenCalled();

    element.unmount();

    expect(onFieldUnmount).toHaveBeenCalled();
  });

  it('should support grouping', () => {
    const model = { testGroup1: { testGroup2: { name: 'value' } } };
    const context: FormContext<any> & FieldGroupContext = {
      ...mockFormContext(model),
      groups: ['testGroup1', 'testGroup2']
    };

    const element = mount(
      <Field name={'name'} render={renderInput}/>
      , { context }
    );
    const input = element.find('input');
    expect(input.props().value).toBe('value');
  });

  it('should support null as name', () => {
    const model = { testGroup1: { testGroup2: 'value' } };
    const context: FormContext<any> & FieldGroupContext = {
      ...mockFormContext(model),
      groups: ['testGroup1', 'testGroup2']
    };

    const element = mount(
      <Field name={null} render={renderInput}/>
      , { context });

    const input = element.find('input');
    expect(input.props().value).toBe('value');
  });

  it('should trigger onFieldChange in group', () => {
    const name = 'name';
    const value = 'value';
    const onFieldChange = jest.fn();
    const field = <Field name={name} render={renderInput}/>;
    const model = { group1: { group2: { [name]: value } } };
    const context: FormContext<any> & FieldGroupContext = {
      ...mockFormContext(model, { onFieldChange }),
      groups: ['group1', 'group2']
    };
    const element = mount(field, { context });
    const wrapper = element.find('input');
    const onChange = wrapper.props().onChange;

    expect(onFieldChange).not.toHaveBeenCalled();

    const newValue = 'newValue';
    onChange!(mockEvent(newValue));

    expect(onFieldChange).toHaveBeenCalledWith(['group1', 'group2', name], newValue);
  });
});
