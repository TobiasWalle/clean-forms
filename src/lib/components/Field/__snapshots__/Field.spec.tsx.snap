// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Field should render 1`] = `
<input
  name="name"
  onBlur={[Function]}
  onChange={[Function]}
  onFocus={[Function]}
  value="value"
/>
`;
